corehq/apps/api/tests/test_case_update_api.py

```diff
@@ -0,0 +1,179 @@
+            'rank': '2100',
+            'sport': 'chess',
+        })
+
+    def test_create_child_case(self):
+        parent_case = self._make_case()
+        res = self.post('create_case', {
+            '@case_type': 'match',
+            '@case_name': 'Harmon/Luchenko',
+            '@owner_id': 'harmon',
+            'properties': {
+                'external_id': '23',
+                'winner': 'Harmon',
+            },
+            'indices': {
+                'parent': {
+                    '@case_id': parent_case.case_id,
+                    '@case_type': 'player',
+                    '@relationship': 'child',
+                },
+            },
+        })
+        self.assertEqual(res.keys(), ['@case_id', '@form_id'])
+
+        case = self.case_accessor.get_case(res['@case_id'])
+        self.assertEqual(case.name, 'Harmon/Luchenko')
+        self.assertEqual(case.owner_id, 'harmon')
+        self.assertEqual(case.dynamic_case_properties(), {
+            'external_id': '23',
+            'winner': 'Harmon',
+        })
+        self.assertEqual(case.indices[0].identifier, 'parent')
+        self.assertEqual(case.indices[0].referenced_id, parent_case.case_id)
+        self.assertEqual(case.indices[0].referenced_type, 'player')
+        self.assertEqual(case.indices[0].relationship_id, 'child')
+        self.assertEqual(case.indices[0].case.case_id, parent_case.case_id)
+
+    def test_bulk_action(self):
+        existing_case = self._make_case()
+        res = self.post('bulk_update', [
+            {
+                # update existing case
+                '@case_id': existing_case.case_id,
+                '@case_name': 'Beth Harmon',
+                '@owner_id': 'us_chess_federation',
+                'properties': {
+                    'rank': '2100',
+                    'champion': 'true',
+                },
+            },
+            {
+                # No case_id means this is a new case
+                '@case_type': 'player',
+                '@case_name': 'Jolene',
+                '@owner_id': 'methuen_home',
+                'properties': {
+                    'external_id': '2',
+                    'sport': 'squash',
+                    'dob': '1947-03-09',
+                },
+            },
+        ])
+        #  only returns a single form ID - chunking should happen in the client
+        self.assertEqual(res.keys(), ['@case_ids', '@form_id'])
+
+        updated_case = self.case_accessor.get_case(existing_case.case_id)
+        self.assertEqual(updated_case.name, 'Beth Harmon')
+
+        new_case = self.case_accessor.get_case(res['@case_ids'][1])
+        self.assertEqual(new_case.name, 'Jolene')
+
+    def test_create_parent_and_child_together(self):
+        # TODO? Since this API doesn't let you provide your own case IDs, you
+        # can't reference uncreated cases in indices unless we invent some
+        # syntax specifically for that. Thinking about leaving out of scope for
```

Comment:
That's a good point.  I'll make a note to try that out.

Reply:
I would also lean away from the idea of nesting cases, in favor of offering both "external_id" and "temporary_id".

In case it's useful to know how other systems do this, this is analogous to FHIR's "Bundle" resource (of type "collection"). To support relationships between resources in a Bundle, [you use references](https://www.hl7.org/fhir/bundle.html#references), which can be internal to the Bundle or an external ID.

Reply:
I like the `temporary_id` idea. If we go down that route, we should namespace it or make sure it wouldn't ever conflict with something else (`_temporary_id` or `metadata -> case_index_id` or something).
I'm always concerned with using `external_id` for things like this as those aren't guaranteed to be unique in a project.